API Availability

mikepie1 opened the pull request:
First draft of section covering the need for the new WebIDL extended
attribute "CheckAnyPermissions"

Commit:
API Availability

First draft of section covering the need for the new WebIDL extended
attribute "CheckAnyPermissions"

Commit:
browser, browserAction updates

Added WebIDL and other info

Commit:
Merge remote-tracking branch 'refs/remotes/browserext/gh-pages' into gh-pages

Commit:
Drafted sections for remaining objects

Drafted sections for remaining objects

Commit:
Spelling fixes, externally_connectable, etc

Did some cleanup, clarification, and addition of
"externally_connectable" manifest tag.

frivoal commented:
Overall, I think this is a good start, and that we should merge this. I have a few comments based on my first read. If you want to address them before merging, great. Otherwise, if you prefer to merge as is, I'm OK with that, and I'll open individual github issues for each comment.
1. A few titles in section 2 use the form "Due to ....". Something like "Restrictions due to ..." or "Restrictions based on ..." or "Availability based on ..." would make it easier to read.
2. Section 2.1.1 (Due to Content Security Policy (CSP)), makes various statements about "... is/are not allowed", or "only ... can be executed". If these are meant to be normative statements, an explanation of what happens if you try anyway is needed. If these are meant as informative overview descriptions further specified later in the spec, a comment to that effect would be welcome. something like 
   
   > _This section is informative. For normative definitions, see section &lt;a href="">foo&lt;/a>_
   > The same comment applies to 2.1.3, which I understand to be normatively defined through IDL later on. I am unsure about 2.1.2 is it normatively defining the various contexts, so that they can be referred to later, or informatively introducing them?
3. "Either browserAction or pageAction may be supported by an extension, not both" What happens if you try to set both?
4.  "CheckAnyPermissions" is defined via a reference to MDN. MDN is great, but it is not a normative source. Should we define it in this spec? Should we ask the the webplatfom WG to put this definition in the IDL spec?
5. "First, let's look at an example of how the browser.contextMenus object would be expressed in WebIDL". I would keep this example, but rephrase the introduction sentence to something like "This section illustrates how the browser.contextMenus, normatively defined in section &lt;a href="...">&lt;/a> is described using IDL and the IDL extensions mentioned above."
6. 2.1.4.4 Example (i18n) / "DISCUSS: It has been suggested that we cut this section." -> I lean towards removing. What does it bring to the table that the previous example doesn't? 

To be continued...

kmaglione commented on index.html:
```diff
@@ -35,83 +46,956 @@
+            </ul>
+        </p><section>
+            <h3 id="apiavailability">API Availability</h3>
+            Browser extensions are authored primarily with JavaScript. In addition to the usual <a href="https://www.w3.org/standards/webdesign/script">JavaScript Web APIs</a>, additional APIs are available exclusively to extensions via the <code>browser</code> object. However, the specific APIs that are available to an extension from the <code>browser</code> object are dependent upon:
```
This should be in a `<p>`

kmaglione commented on index.html:
```diff
@@ -35,83 +46,956 @@
+        <h2>Core APIs</h2>
+        <p>
+            The goal for these core APIs is to provide functionality that extension authors need for:<br />
+            <ul>
```
The `<ul>` should not be a child of a `<p>`

kmaglione commented on index.html:
```diff
@@ -35,83 +46,956 @@
+            <section>
+                <h4>Due to Manifest Keys and Permissions</h4>
+                Some browser extension API require a specific manifest key or specific permission. If the key or permission are not declared or not granted, the object will not be available for scripting. These requirements are summarized in the table below.<br /><br />
+                <font style="background:yellow">DISCUSS: Should we remove the "Optional Permissions" column? Do we plan to support this in the common/core spec?</font><br/>
```
Please use `<span class="...">` rather than `<font>` tags, and `<p>` rather than `<br>`

frivoal replied:
Please use `class="issue"`, that's what it's for.

kmaglione commented on index.html:
```diff
@@ -35,83 +46,956 @@
+[NoInterfaceObject]
+interface BrowserContextMenus {
+    unsigned short create(ContextMenuCreateProperties data, Function callback);
+    object onClicked(Function callback);
```
We're going to have to figure out how to handle event listeners. This describes a function. For the short term, we should probably go with something generic like `readonly attribute EventListener onClicked;`, and define the `EventListener` type.

kmaglione commented on index.html:
```diff
@@ -35,83 +46,956 @@
+                    <font style="background:yellow;">DISCUSS: It has been suggested that we cut these examples.</font><br />
+                    First, let's look at an example of how the <code>browser.contextMenus</code> object would be expressed in WebIDL. In this case, the <code>contextMenus</code> object is only available to the browser host contexts such as background, vent, popup, and options pages. Also, the <code>contextMenus</code> object requires the extension author to specify <code>"permissions":["tabs"]</c ...
+                    <br />
+                    1) First, we declare the methods for the <code>browser.contextMenus</code> object.<br />
```
This should be an `<ol>`

kmaglione commented on index.html:
```diff
@@ -35,83 +46,956 @@
+  void    removeListener(Function callback);
+  boolean hasListener(Function callback);
+  boolean hasListeners();
+  void    addRules(Rules rules);
```
Hm. Where did `Rule`s come from?

kmaglione commented on index.html:
```diff
@@ -35,83 +46,956 @@
+            <code>addListener()</code>.<br />
+            <font style="background:yellow;">TODO: Need to define callbacks throughout this spec, rather than generic "Function callback".</font><br />
+<pre class="def idl">
+[Exposed=(Window,ContentScript) CheckAnyPermissions="browserextensions"]
```
This should probably be `NoInterfaceObject` rather than `Exposed` and `CheckAnyPermissions`. And there need to be commas between the attributes.

kmaglione commented on index.html:
```diff
@@ -35,83 +46,956 @@
+
+        <section>
+            <h3 id="coreapis">The <code>browser</code> object</h3>
+            <b>Overview</b><br />
```
These should be `<h4>`s, or similar, rather than `<b>`

kmaglione commented on index.html:
```diff
@@ -35,83 +46,956 @@
+typedef sequence&lt;BrowserExtensionIcon&gt; BrowserExtensionIconArray;
+dictionary Developer {
+  DOMString name;
+  optional DOMString? url;
```
Remove `optional`

kmaglione commented:
I'll have more comments tomorrow.

Commit:
Addressed feedback from 2016.09.08

Got read of faux headings with <b> and now use <h4
class="introductory">. Cleaned up nesting of <ul> and <p>. Replaced
<br/> with <p>. Replaced <font> with <p class="issue">. Added commas
before CheckAnyPermissions extended attributes in ExtensionEvents. Added
NoInterfaceObject to ExtensionEvents. Got rid of Rules in
BrowserExtensionEvent. Got rid of ContextMenus example and replaced "1)"
with <ol> in i18n example. Since "optional" is only valid for
parameters, I removed "optional" from dictionaries. I have not yet
addressed feeback from Kris about event listeners (e.g. readonly
attribute EventListener onClicked). Next I plan to add Callback
signatures.
